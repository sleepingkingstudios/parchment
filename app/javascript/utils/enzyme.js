import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
