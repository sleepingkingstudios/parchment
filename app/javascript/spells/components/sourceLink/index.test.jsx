import React from 'react';
import { shallow } from 'enzyme';

import SpellSourceLink from './index';

describe('<SpellSourceLink />', () => {
  const defaultProps = {};

  it('should return a message', () => {
    const rendered = shallow(<SpellSourceLink {...defaultProps} />);

    expect(rendered).toHaveDisplayName('span');
    expect(rendered).toHaveClassName('text-muted');
    expect(rendered).toHaveProp('children', 'Homebrew');
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<SpellSourceLink {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('with source: value and sourceType: value', () => {
    const source = {
      id: '00000000-0000-0000-0000-000000000000',
      name: 'sprocket',
    };
    const sourceType = 'Widget';
    const props = { ...defaultProps, source, sourceType };

    it('should return a link', () => {
      const rendered = shallow(<SpellSourceLink {...props} />);
      const { name } = source;
      const url = `/widgets/${source.id}`;

      expect(rendered).toHaveDisplayName('Link');
      expect(rendered).toHaveProp('to', url);
      expect(rendered).toHaveProp('children', name);
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<SpellSourceLink {...props} />);

      expect(rendered).toMatchSnapshot();
    });
  });
});
