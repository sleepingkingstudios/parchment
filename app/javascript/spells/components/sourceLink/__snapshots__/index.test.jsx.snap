// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpellSourceLink /> should match the snapshot 1`] = `
<span
  className="text-muted"
>
  Homebrew
</span>
`;

exports[`<SpellSourceLink /> with source: value and sourceType: value should match the snapshot 1`] = `
<Link
  to="/widgets/00000000-0000-0000-0000-000000000000"
>
  sprocket
</Link>
`;
